netket/sampler/autoreg.py

```diff
@@ -157,7 +157,7 @@ def scan_fun(carry, index):
         # We just need a buffer for `σ` before generating each sample
         # The result does not depend on the initial contents in it
         σ = jnp.zeros(
-            (chain_length * sampler.n_chains_per_rank, sampler.hilbert.size),
+            (sampler.n_chains_per_rank * chain_length, sampler.hilbert.size),
```

Comment:
maybe at some point we should also change the APIs of the sampler to make them agnostic to MPI logic, for example eliminating n_chains_per_rank etc and leaving only total_samples or similar ?

Reply:
I think there is no reason to keep distinguishing samples coming from different ranks in addition to already distinguishing different chains per rank. we should just make sure we know what chains are independent, no matter where they are physically stored, moving a bit closer to what we need for `pmap` sharding etc

Reply:
The main reason we used `n_chains_per_rank` was to make it easier to follow 'weak scaling' instead of 'hard scaling' arguments when doing simulations over multiple nodes. For example with GPUs it makes often no sense to use fewer than 250 chains per gpu so you'd want to use `n_chains_per_rank=250`. 

I think it's handy.

Besides we only actually store one of the two numbers (I think `n_chains_per_rank`) and compute `n_chains` from it. 
I think we could rename this to `n_chains_per_device` (because in any case if you use pmap you will have to take `n_chains` to be a multiple of the number of devices, and then make it work correctly with pmap. 

So instead of having `n_chains = n_chains_per_device * mpi.n_ranks ` we could multiply it by `jax.n_devices()`. we just need some logic to switch from one to the other..

Reply:
@inailuig we agree that this line changes nothing? I think integer multiplication is commutative...

Reply:
@inailuig

Reply:
` _chains_per_rank` is the number of chains jax knows about (and unless you you use mpi its equal `n_chains`),  so that is the number of chains we have to use for creating the arrays.
 
I propose we discuss changing the n_chains logic in the next PR.